ocw_data_parser/course_downloader.py

```diff
@@ -0,0 +1,69 @@
+import os
+import shutil
+import json
+import boto3
+
+"""
+This is a class used for downloading source json from S3 based on a list of course id's
+
+An example of the expected format can be found in example_courses.json
+"""
+class OCWDownloader(object):
+    def __init__(self,
+                 courses_json="",
+                 env="QA",
+                 destination_dir="",
+                 s3_bucket_name="",
+                 overwrite=False):
+        self.coursesJson = courses_json
+        self.env = env
+        self.destinationDir = destination_dir
+        self.s3BucketName = s3_bucket_name
+        self.overwrite = overwrite
+
+    def downloadCourses(self):
+        courses = None
+        with open(self.coursesJson) as f:
+            courses = json.load(f)["courses"]
+        if not os.path.exists(self.destinationDir):
+            os.makedirs(self.destinationDir)
+        s3 = boto3.resource("s3")
+        s3Client = boto3.client("s3")
+        ocwContentStorage = s3.Bucket(self.s3BucketName)
+        allObjects = ocwContentStorage.meta.client.list_objects(
+            Bucket=ocwContentStorage.name, Prefix="{}/".format(self.env), Delimiter="/")
+
+        # first tier subfolders under env should be department numbers
+        for departmentPath in allObjects.get("CommonPrefixes"):
+            # second tier subfolders should be course numbers
+            courseNumbers = ocwContentStorage.meta.client.list_objects(
+                Bucket=ocwContentStorage.name, Prefix=departmentPath.get("Prefix"), Delimiter="/")
+            for courseNumberPath in courseNumbers.get("CommonPrefixes"):
+                # third tier folders should be course terms
+                courseTerms = ocwContentStorage.meta.client.list_objects(
+                    Bucket=ocwContentStorage.name, Prefix=courseNumberPath.get("Prefix"), Delimiter="/")
+                for courseTerm in courseTerms.get("CommonPrefixes"):
+                    # fourth tier folders should be course folders keyed with the course id
+                    courseFolders = ocwContentStorage.meta.client.list_objects(
+                        Bucket=ocwContentStorage.name, Prefix=courseTerm.get("Prefix"), Delimiter="/")
+                    for courseFolder in courseFolders.get("CommonPrefixes"):
+                        # make sure the course is in our list
+                        courseId = os.path.basename(
+                            os.path.normpath(courseFolder.get("Prefix")))
+                        if courseId in courses:
```

Comment:
Is course id something along the lines on "12-001-introduction-to-geology-fall-2013". Should it be course_file_name instead?

Reply:
In the JSON files, this string is called "course_identifier" so I thought it was appropriate.